### PR TITLE
LibCore: Add CDirIterator, and use it in everything rather than readdir

### PR DESCRIPTION
**Pull request by rburchell**

### CODE REVIEW

File: LibCore/CDirIterator.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        errno = 0;
+        auto* de = readdir(m_dir);
+        if (de) {
+            m_next = de->d_name;
+        } else {
+            m_error = errno;
+            m_next = String();
+        }
+
+        if (m_next == String()) {
```
Comment (awesomekling) [inline LibCore/CDirIterator.cpp]: Let's say `m_next.is_null()`. You're already doing that elsewhere :)

File: LibCore/CDirIterator.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    return advance_next();
+}
+
+String CDirIterator::next_path()
+{
+    if (m_next.is_null())
+        advance_next();
+
+    auto tmp = m_next;
```
Comment (awesomekling) [inline LibCore/CDirIterator.cpp]: Oh, maybe we should add an AK::exchange() like std::exchange()

File: LibCore/CDirIterator.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    ~CDirIterator();
+
+    bool has_error() const { return m_error != 0; }
+    int error() const { return m_error; }
+    const char *error_string() const { return strerror(m_error); }
+    bool has_next();
+    String next_path();
+
+private:
+    DIR *m_dir = nullptr;
```
Comment (awesomekling) [inline LibCore/CDirIterator.h]:
```suggestion
    DIR* m_dir = nullptr;
```

File: LibCore/CDirIterator.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        NoFlags = 0x0,
+        SkipDots = 0x1,
+    };
+
+    CDirIterator(const String& path, Flags flags = Flags::NoFlags);
+    ~CDirIterator();
+
+    bool has_error() const { return m_error != 0; }
+    int error() const { return m_error; }
+    const char *error_string() const { return strerror(m_error); }
```
Comment (awesomekling) [inline LibCore/CDirIterator.h]:
```suggestion
    const char* error_string() const { return strerror(m_error); }
```

File: LibCore/CDirIterator.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#include <AK/AKString.h>
+
+class CDirIterator {
+public:
+    enum Flags {
+        NoFlags = 0x0,
+        SkipDots = 0x1,
+    };
+
+    CDirIterator(const String& path, Flags flags = Flags::NoFlags);
```
Comment (awesomekling) [inline LibCore/CDirIterator.h]:
```suggestion
    CDirIterator(const String& path, Flags = Flags::NoFlags);
```

File: LibCore/CDirIterator.cpp (hunk 2 above)
Comment (rburchell) [reply LibCore/CDirIterator.cpp]: I actually wasn't familiar with it. But that sounds like a reasonable idea.

**Comment by awesomekling** [comment]
Delicious :)

**Pull request merged**